pumpkin-lib/src/engine/sat/restart_strategy.rs

```diff
@@ -189,56 +189,69 @@ impl RestartStrategy {
         // average lbd multiplied by the `lbd_coefficient` is lower than the short-term average lbd
         self.lbd_long_term_moving_average.value() * self.lbd_coefficient
             <= self.lbd_short_term_moving_average.value()
     }
 
+    fn should_restart_first_time(&self) -> bool {
+        self.number_of_restarts == 0
+            && self.number_of_conflicts_encountered_since_restart
+                < self.minimum_number_of_conflicts_before_first_restart
+    }
+
     /// Notifies the restart strategy that a conflict has taken place so that it can adjust its
     /// internal values, this method has the additional responsibility of checking whether a restart
     /// should be blocked based on whether the solver is "sufficiently close" to finding a solution.
-    pub(crate) fn notify_conflict(&mut self, lbd: u32, num_literals_on_trail: usize) {
+    pub(crate) fn notify_conflict(&mut self, lbd: u32, number_of_pruned_values: u64) {
         // Update moving averages
         self.number_of_assigned_variables_moving_average
-            .add_term(num_literals_on_trail as u64);
+            .add_term(number_of_pruned_values);
         self.lbd_short_term_moving_average.add_term(lbd as u64);
         self.lbd_long_term_moving_average.add_term(lbd as u64);
 
         // Increase the number of conflicts encountered since the last restart
         self.number_of_conflicts_encountered_since_restart += 1;
 
-        // If the solver has more variables assigned now than in the recent past, then block the
-        // restart. The idea is that the solver is 'closer' to finding a solution and restarting
-        // could be harmful to the performance
-        if (self.number_of_restarts > 0
-            || self.number_of_conflicts_encountered_since_restart
-                >= self.minimum_number_of_conflicts_before_first_restart)
-            && self.number_of_conflicts_until_restart
-                <= self.number_of_conflicts_encountered_since_restart
-            && num_literals_on_trail as f64
-                > self.number_of_assigned_variables_moving_average.value()
-                    * self.number_of_variables_coefficient
-        {
+        if self.should_block_restart(number_of_pruned_values) {
             // Restart has been blocked
             self.number_of_blocked_restarts += 1;
             self.reset_values()
         }
     }
 
+    fn should_block_restart(&self, number_of_pruned_values: u64) -> bool {
+        // If the solver has more variables assigned now than in the recent past, then block the
+        // restart. The idea is that the solver is 'closer' to finding a solution and restarting
+        // could be harmful to the performance
+
+        if self.should_restart_first_time() {
+            // Don't block the first restart.
+            return false;
+        }
+
+        let close_to_solution = number_of_pruned_values as f64
```

Comment:
(not for this PR) later we could investigate different ways of making this estimate!